### PR TITLE
Optimize array_equivalent for NDFrame.equals

### LINKED ISSUE #35249: Performance Regression in frame_methods.Equals.time_frame_float_equal
https://pandas.pydata.org/speed/pandas/index.html#frame_methods.Equals.time_frame_float_equal?commits=77f4a4d673e2d2b3f7396a94818771d332bd913c

Some discussion starting at https://github.com/pandas-dev/pandas/pull/34962#issuecomment-655456601

@jbrockmendel what needs to be done for 1.1, just optimize `array_equivalent` or something larger? Is this a blocker (if so can you  add the blocker tag)?

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes #35249

```python
In [1]: import numpy as np
In [2]: import pandas as pd
In [3]: N = 10**3
In [4]: float_df = pd.DataFrame(np.random.randn(N, N))
# 1.0.5
1.52 ms ± 94 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

# master
83.6 ms ± 1.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# branch
31.2 ms ± 2.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

And proof of this being better for short-circuiting:

```python
In [4]: N = 10**3

In [5]: a = pd.DataFrame(np.random.randn(N, N))

In [6]: b = pd.DataFrame(np.random.randn(N, N))

In [7]: %timeit a.equals(b)
# 1.0.5
1.97 ms ± 127 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [10]: %timeit a.equals(b)
56.4 µs ± 1.67 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Discussion: the key thing here is that `BlockManager.equals` requires that the dtypes of `left` and `right` match. If they don't then we've already returned False. That lets us do a few things

1. Avoid many `is_foo_dtype` checks.
2. Use optimized checks like `np.isnan`
3. Avoid the slow case of considering `np.array([1, 2])` equivalent to `np.array([1.0, 2.0])`.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
restarting failed travis job 

```
=========================== short test summary info ============================
FAILED pandas/tests/io/parser/test_common.py::test_chunks_have_consistent_numerical_type[python]
= 1 failed, 69943 passed, 4255 skipped, 1020 xfailed, 13 warnings in 808.93s (0:13:28) =
```

File: pandas/core/dtypes/missing.py [hunk 1]
```diff
@@ -391,43 +399,28 @@ def array_equivalent(left, right, strict_nan: bool = False) -> bool:
     if left.shape != right.shape:
         return False
 
+    if dtype_equal:
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:402]: i am not sure i understand why you actually need this
the dtype check is cheap compared to the actual check no?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/dtypes/missing.py:402]: 1. They add up, especially when doing things columnwise for small, wide dataframes
2. We get to skip a few other checks, like swapping `np.isnan` for `pd.isna` for float, and we can skip the empty check at https://github.com/pandas-dev/pandas/pull/35328/files#diff-ff8364cee9a3e1ef3a3825cb2cdd26d8L431.

Comment (jreback) [reply pandas/core/dtypes/missing.py:402]: ok that's fine

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -452,6 +445,45 @@ def array_equivalent(left, right, strict_nan: bool = False) -> bool:
     return np.array_equal(left, right)
 
 
+def array_equivalent_float(left, right):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:448]: maybe make all of these helpers private functions

**Review by jreback — APPROVED** [review]

File: pandas/core/dtypes/missing.py [hunk 3]
```diff
@@ -452,6 +445,45 @@ def array_equivalent(left, right, strict_nan: bool = False) -> bool:
     return np.array_equal(left, right)
 
 
+def _array_equivalent_float(left, right):
+    return ((left == right) | (np.isnan(left) & np.isnan(right))).all()
```
Comment (jbrockmendel) [inline pandas/core/dtypes/missing.py:449]: what about use_inf_as_na?

Can this be re-used on L426?  (also the prod(shape) check on L424-425 i think is redundant with earlier shape check)

Comment (TomAugspurger) [reply pandas/core/dtypes/missing.py:449]: It seems like 1.0.5 doesn't care about that

```python
In [15]: df1 = pd.DataFrame({"A": np.array([np.nan, 1, np.inf])})

In [16]: df2 = pd.DataFrame({"A": np.array([np.nan, 1, np.nan])})

In [17]: with pd.option_context('mode.use_inf_as_na', True):
    ...:     print(df1.equals(df2))
    ...:
    ...:
False
```

**Comment by TomAugspurger** [comment]
Travis finished. I'm going to selfishly ignore https://github.com/pandas-dev/pandas/pull/35328#discussion_r456589372 so that I can get the RC process rolling, but happy to engage more on followups if I missed something.

File: pandas/core/dtypes/missing.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py:449]: Makes sense, thanks for checking

**Pull request merged**